karapace/protobuf/proto_normalizations.py

```diff
@@ -0,0 +1,220 @@
+        options=sorted_options,
+        reserveds=message_element.reserveds,
+        fields=sorted_fields,
+        one_ofs=sorted_one_ofs,
+        extensions=message_element.extensions,
+        groups=message_element.groups,
+    )
+
+
+def type_element_with_sorted_options(type_element: TypeElement) -> TypeElement:
+    sorted_neasted_types: List[TypeElement] = []
+
+    for nested_type in type_element.nested_types:
+        if isinstance(nested_type, EnumElement):
+            sorted_neasted_types.append(enum_element_with_sorted_options(nested_type))
+        elif isinstance(nested_type, MessageElement):
+            sorted_neasted_types.append(message_element_with_sorted_options(nested_type))
+        else:
+            raise ValueError("Unknown type element")  # tried with assert_never but it did not work
+
+    # doing it here since the subtypes do not declare the nested_types property
+    type_element.nested_types = sorted_neasted_types
+
+    if isinstance(type_element, EnumElement):
+        return enum_element_with_sorted_options(type_element)
+
+    if isinstance(type_element, MessageElement):
+        return message_element_with_sorted_options(type_element)
+
+    raise ValueError("Unknown type element")  # tried with assert_never but it did not work
+
+
+def extends_element_with_sorted_options(extend_element: ExtendElement) -> ExtendElement:
+    sorted_fields = (
+        None
+        if extend_element.fields is None
+        else [type_field_element_with_sorted_options(field) for field in extend_element.fields]
+    )
+    return ExtendElement(
+        location=extend_element.location,
+        name=extend_element.name,
+        documentation=extend_element.documentation,
+        fields=sorted_fields,
+    )
+
+
+def rpc_element_with_sorted_options(rpc: RpcElement) -> RpcElement:
+    sorted_options = None if rpc.options is None else list(sorted(rpc.options, key=sort_by_name))
+    return RpcElement(
+        location=rpc.location,
+        name=rpc.name,
+        documentation=rpc.documentation,
+        request_type=rpc.request_type,
+        response_type=rpc.response_type,
+        request_streaming=rpc.request_streaming,
+        response_streaming=rpc.response_streaming,
+        options=sorted_options,
+    )
+
+
+def service_element_with_sorted_options(service_element: ServiceElement) -> ServiceElement:
+    sorted_options = None if service_element.options is None else list(sorted(service_element.options, key=sort_by_name))
+    sorted_rpc = (
+        None if service_element.rpcs is None else [rpc_element_with_sorted_options(rpc) for rpc in service_element.rpcs]
+    )
+
+    return ServiceElement(
+        location=service_element.location,
+        name=service_element.name,
+        documentation=service_element.documentation,
+        rpcs=sorted_rpc,
+        options=sorted_options,
+    )
+
+
+def normalize_options_ordered(proto_file_element: ProtoFileElement) -> ProtoFileElement:
```

Comment:
This can be simply named as `normalize`, no need for different kind of normalizations?

But should it be method of ProtobufSchema or ProtoFileElement instead of separate function?

Reply:
initially I was considering of implementing the different types of normalization into different prs and with different goals, like normalizing the option orders, normalize the string names etc etc. The normalize function would be much more harder than this one and I thought about separating the tests and the guarantees of those phases. Probably isn't a good idea though :)